### PR TITLE
Apply the early check of the universe declaration and the monomorphic universe fixing to all interactive declarations

### LINKED ISSUE #19289: Incorrect "universe is unbound" error in early check for Lemma (missing restrict)
~~~coq
Set Universe Polymorphism.

Fail Lemma foo@{} : (nat :> Type).
(* universe is unbound *)

Lemma foo : (nat :> Type).
Proof. exact 0. Qed.

Check foo@{}.
~~~
(same error without univ poly but then we can't as easily check that the universe can be restricted away)

### PR DESCRIPTION
**Pull request by herbelin**
Formerly, the early check (at `Definition` time) of the universe declaration and the monomorphic universe fixing (for asynchronous evaluation) were done only for `Theorem` and `Definition` without body (in `post_check_evd`). We do it also for interactive [`Co`]`Fixpoint` and for `Program Fixpoint` and `Program Definition`.

To anticipate what is done when proofs are completed, we apply `UState.collapse_sort_variables` and `UState.restrict` in some cases (this is needed for the test-suite to pass). But maybe should these be applied in all four cases, together with `UState.normalize_variables` and `UState.minimize`, as done at the end of proofs in general???

This gives an alternative solution to #15410 which allows to remove the forbidding for extensible universes introduced in #15424.

This remark at #15410 still hold though: "non extensible universe declarations don't make a lot of sense in universe monomorphic definitions", so maybe it can also be a failure.

- [x] Added / updated **test-suite**.
- [x] TODO: use the same check in all four situations (but which one?)
- [x] Added change log
 
Depends on #18915 (merged)

Fix #19289

### CODE REVIEW

**Comment by herbelin** [comment]
In two cases, `comDefinition.ml` and `comProgramFixpoint.ml`, I apparently need to do:
```ocaml
let check_univ_decl_early ~poly sigma udecl terms =
let vars = List.fold_left (fun acc b -> Univ.Level.Set.union acc (Vars.universes_of_constr b)) Univ.Level.Set.empty terms in
let uctx = Evd.evar_universe_context sigma in
let uctx = UState.collapse_sort_variables uctx in
let uctx = UState.restrict uctx vars in
UState.check_univ_decl ~poly uctx udecl
```
But in `comFixpoint.ml` and `vernacentries.ml`, `collapse_sort_variable` is not needed (for the examples of the test-suite to work). Is the need for `collapse_sort_variable` really expected or is it a revelator of a possible problem elsewhere?

**Comment by herbelin** [comment]
I'm asking because I found sometimes free qsort in the absence of minimization. Now, I also realize that `vernacentries.ml` and `comFixpoint.ml` do have a minimization and the others not. So, I have the feeling that what is finally needed is a minimization. I will try, thanks for the help.

**Comment by herbelin** [comment]
Finally, I used collapse_sort_variable since I saw it is already called by minimization and that, since the result is thrown away after checking, an (a priori) faster function is better.

File: test-suite/bugs/bug_15410.v [hunk 1]
```diff
@@ -1,3 +1,4 @@
 Require Program.Tactics.
 
-Fail Program Definition foo@{u} : Type@{u} := _.
+Program Definition foo@{u} : Type@{u} := _.
+Next Obligation. exact nat. Qed.
```
Comment (SkySkimmer) [inline test-suite/bugs/bug_15410.v:4]: What happens if we do `exact Type` instead? The univ decl is non extensible so it should be an error.

Comment (herbelin) [reply test-suite/bugs/bug_15410.v:4]: > What happens if we do `exact Type` instead? 

It succeeds in the PR, because the extra monomorphic universe is not in `foo` but in the obligation `foo_obligation_1`. Should the universe declaration of a `Program` definition include the monomorphic universes of the obligations? Maybe?

Comment (SkySkimmer) [reply test-suite/bugs/bug_15410.v:4]: this question is basically why I decided that monomorphic non extensible univ declarations are not really sensible

Comment (herbelin) [reply test-suite/bugs/bug_15410.v:4]: OK, I see. So, to support it, we would either need to invent a syntax to pre-annotate the universes of all obligations (complicated), or to collect the monomorphic universes of the obligations and do an ultimate check (as done for `Theorem` to reject the definition). Or we keep the requirement of extensibility for `Program`, which is quite easy to implement.

Comment (herbelin) [reply test-suite/bugs/bug_15410.v]: I just pushed an extra commit that requires all kinds of `Program`-based definitions to have extensible universe declarations (and reports an error otherwise, as currently done in master).

**Comment by herbelin** [comment]
Am I correct that the CI failures are unrelated?

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: test-suite/bugs/bug_15410.v [hunk 2]
```diff
@@ -1,3 +1,11 @@
-Require Program.Tactics.
+Require Program.
```
Comment (SkySkimmer) [inline test-suite/bugs/bug_15410.v:1]: Why this change?

Comment (herbelin) [reply test-suite/bugs/bug_15410.v:1]: Not necessary indeed. Will wait for CI completion before removing the change.

Comment (SkySkimmer) [reply test-suite/bugs/bug_15410.v:1]: ping

**Comment by SkySkimmer** [comment]
Is the coq_lsp failure real?

**Comment by ppedrot** [comment]
This is a missing overlay, I think.

**Comment by SkySkimmer** [comment]
@coqbot bench

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now
